crates/gen-guest-teavm-java/src/lib.rs

```diff
@@ -0,0 +1,1918 @@
+                results: lowered_types,
+                ..
+            } => self.lower_variant(
+                &union
+                    .cases
+                    .iter()
+                    .map(|case| Some(case.ty))
+                    .collect::<Vec<_>>(),
+                lowered_types,
+                iface,
+                &operands[0],
+                results,
+            ),
+
+            Instruction::UnionLift { union, ty, .. } => {
+                let cases = union
+                    .cases
+                    .iter()
+                    .enumerate()
+                    .map(|(i, case)| (format!("f{i}"), case.ty))
+                    .collect::<Vec<_>>();
+
+                self.lift_variant(
+                    &Type::Id(*ty),
+                    &cases
+                        .iter()
+                        .map(|(name, ty)| (name.deref(), Some(*ty)))
+                        .collect::<Vec<_>>(),
+                    iface,
+                    &operands[0],
+                    results,
+                )
+            }
+
+            Instruction::OptionLower {
+                results: lowered_types,
+                payload,
+                ..
+            } => self.lower_variant(
+                &[None, Some(**payload)],
+                lowered_types,
+                iface,
+                &operands[0],
+                results,
+            ),
+
+            Instruction::OptionLift { payload, ty } => self.lift_variant(
+                &Type::Id(*ty),
+                &[("none", None), ("some", Some(**payload))],
+                iface,
+                &operands[0],
+                results,
+            ),
+
+            Instruction::ResultLower {
+                results: lowered_types,
+                result,
+                ..
+            } => self.lower_variant(
+                &[result.ok, result.err],
+                lowered_types,
+                iface,
+                &operands[0],
+                results,
+            ),
+
+            Instruction::ResultLift { result, ty } => self.lift_variant(
+                &Type::Id(*ty),
+                &[("ok", result.ok), ("err", result.err)],
+                iface,
+                &operands[0],
+                results,
+            ),
+
+            Instruction::EnumLower { .. } => results.push(format!("{}.ordinal()", operands[0])),
+
+            Instruction::EnumLift { name, .. } => results.push(format!(
+                "{}.values()[{}]",
+                name.to_upper_camel_case(),
+                operands[0]
+            )),
+
+            Instruction::ListCanonLower { element, realloc } => {
+                let op = &operands[0];
+                let (size, ty) = list_element_info(element);
+
+                // Note that we can only reliably use `Address.ofData` for elements with alignment <= 4 because as
+                // of this writing TeaVM does not guarantee 64 bit items are aligned on 8 byte boundaries.
```

Comment:
We've had discussions historically about alignment in the canonical ABI and whether or not it should be required and traps if invalid. If this is too onerous it may be worth reopening the discussions to specify alignment and not require it like core wasm does.

Reply:
I wouldn't say it's onerous.  If it turns out the extra copies for `list<u64>` values create real-world performance problems we can either revisit that discussion or I can see about modifying TeaVM to align `long[]` as desired.